cmd/proxy-server/main.go

```diff
@@ -540,3 +563,152 @@ func rolesHandler(log *logrus.Entry, opaHost string) http.Handler {
+
+		authz := r.Header.Get("Authorization")
+		parts := strings.Split(authz, " ")
+		if len(parts) != 2 {
+			log.Println("invalid authz header")
+			return
+		}
+		scheme, tkn := parts[0], parts[1]
+
+		switch scheme {
+		case "Bearer":
+			var claims token.Claims
+			//if token can is valid
+			_, err := tm.ParseWithClaims(tkn, JWTSigningSecret, &claims)
+			if err != nil {
+				w.WriteHeader(http.StatusUnauthorized)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			// Check if the tenant is being denied.
+			ok, err := rdb.SIsMember(keyTenantRevoked, claims.Group).Result()
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			if ok {
+				w.WriteHeader(http.StatusUnauthorized)
+				if err := web.JSONErrorResponse(w, fmt.Errorf("tenant is revoked")); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			log.Info("Calling into roles!")
+
+			// Gather storage pool, systemtype and systemID
+
+			if roleServ.roleService == nil {
+				resp, err = roleServ.roleClient.List(r.Context(), &pb.RoleListRequest{})
+			} else {
+				resp, err = roleServ.roleService.List(r.Context(), &pb.RoleListRequest{})
+			}
+
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+
+			roleJSON := roles.NewJSON()
+			err = roleJSON.UnmarshalJSON(resp.Roles)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+
+			matches := []roles.Instance{}
+			rolesSplit := strings.Split(claims.Roles, ",")
+
+			roleJSON.Select(func(rInst roles.Instance) {
+				for _, role := range rolesSplit {
+					if rInst.Name == role {
+						matches = append(matches, rInst)
+						log.Info("found role!")
+						sysID = rInst.SystemID
+						storPool = rInst.Pool
+						sysType = rInst.SystemType
+						tenant = claims.Group
+
+						dataKey := fmt.Sprintf("quota:%s:%s:%s:%s:data", sysType, sysID, storPool, tenant)
+
+						res, err := rdb.HGetAll(dataKey).Result()
+						if err != nil || len(res) == 0 {
+							w.WriteHeader(http.StatusInternalServerError)
+							if err := web.JSONErrorResponse(w, fmt.Errorf("no volumes found for tenant %s", tenant)); err != nil {
+								log.WithError(err).Println("sending json response")
+							}
+							return
+						}
+
+						for volKey := range res {
+							if strings.Contains(volKey, "capacity") {
+								splitStr := strings.Split(volKey, ":")
+								//example : vol:k8s-cb89d36285:capacity
+								if len(splitStr) == 3 {
```

Comment:
If the length of the split is not 3, what should happen?

Reply:
We skip the raw without vol:vo-name:approved